### PR TITLE
 Add GitHub action to format and lint code

### PR DESCRIPTION
**Pull request by dyastremsky**
This pull request adds a pre-commit to run the following:
- Black for formatting Python
- Flake8 for linting Python for conformance to PEP8, PyFlakes, and circular complexity (max-line-length set to match Black's 88 default/recommended character limit; this and other settings from Black's documentation [here](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-length) to avoid enforcement of limit when unnecessary).
- isort for deterministically sorting and organizing Python imports
- Clang for formatting C++, C, Java, TextProto, Proto, and CUDA
- Codespell for finding common spelling mistakes 

In addition, this adds these native pre-commit hooks, details [here](https://pre-commit.com/hooks.html):
- check-case-conflict
- check-executables-have-shebangs
- check-merge-conflict
- check-json
- check-toml
- check-yaml
- check-shebang-scripts-are-executable
- end-of-file-fixer
- mixed-line-ending
- requirements-txt-fixer
- trailing-whitespace

To run these locally, you can go into the repo and run `pip install pre-commit` and `pre-commit install`. After that has been done once, you just need to run `pre-commit run --all-files` any time you want to apply them. Once installed, it should run automatically on commit for files included in a commit. To correct spelling errors found by codespell when possible, you can install codespell (`pip install codespell` and potentially `pip install --upgrade codespell` to use the toml file in the directory) and call `codespell -w <path>`.

This pull request also adds a GitHub action so that these are run for every pull request and push to main. The core changes are in pre-commit.yml, pre-commit-config.yaml, and pyproject.toml. The rest of the changes are the effect of applying the action to the current repo.

### CODE REVIEW

File: .github/workflows/pre-commit.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+name: pre-commit
+
+on:
+  pull_request:
+  push:
+    branches: [main]
+
+jobs:
+  pre-commit:
+    runs-on: ubuntu-latest
```
Comment (tanmayv25) [inline .github/workflows/pre-commit.yml:36]:
```suggestion
    runs-on: ubuntu-22.04
```

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,62 @@
+# Copyright 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (tanmayv25) [inline pyproject.toml:1]: The copyright seems to be repeated...

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,92 @@
+# Copyright 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (tanmayv25) [inline .pre-commit-config.yaml:1]: Repeated copyrights.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+- repo: https://github.com/codespell-project/codespell
+  rev: v2.2.4
+  hooks:
+      - id: codespell
+        additional_dependencies: [tomli]
+        args: ["--toml", "pyproject.toml"]
+        exclude: (?x)^(.*stemmer.*|.*stop_words.*|^CHANGELOG.md$)
+- repo: https://github.com/pre-commit/pre-commit-hooks
+  rev: v4.4.0
+  hooks:
```
Comment (tanmayv25) [inline .pre-commit-config.yaml:85]: Can we add a comment here with this link: https://pre-commit.com/hooks.html
for more details about these hooks?

File: pyproject.toml (hunk 2 above)
Comment (dyastremsky) [reply pyproject.toml:1]: Thanks for catching this, likely during a copy and paste. Fixed in all PRs.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (dyastremsky) [reply .pre-commit-config.yaml:1]: Fixed.

File: .pre-commit-config.yaml (hunk 4 above)
Comment (dyastremsky) [reply .pre-commit-config.yaml:61]: Sure, done.

**Review by Tabrizian — APPROVED** [review]

**Pull request merged**